README.markdown
# Apache HTTP 2.4 - Docker image for SSL/TLS Mutual Authentication
[![Antonio Musarra's Blog](https://img.shields.io/badge/maintainer-Antonio_Musarra's_Blog-purple.svg?colorB=6e60cc)](https://www.dontesta.it)
[![Build Status](https://travis-ci.org/amusarra/docker-apache-ssl-tls-mutual-authentication.svg?branch=master)](https://travis-ci.org/amusarra/docker-apache-ssl-tls-mutual-authentication)
[![Twitter Follow](https://img.shields.io/twitter/follow/antonio_musarra.svg?style=social&label=%40antonio_musarra%20on%20Twitter&style=plastic)](https://twitter.com/antonio_musarra)

L'obiettivo di questo progetto è quello di fornire un **template** pronto all'uso
e che realizza un sistema di **mutua autenticazione o autenticazione bilaterale SSL/TLS** 
basato su [Apache HTTP](http://httpd.apache.org/docs/2.4/). 
**Ognuno è libero poi di modificare o specializzare questo progetto sulla base delle proprie esigenze**

Se sei impaziente allora potresti provare con Play-With-Docker

[![Try in PWD](https://raw.githubusercontent.com/play-with-docker/stacks/master/assets/images/button.png)](https://labs.play-with-docker.com/?stack=https://raw.githubusercontent.com/amusarra/docker-apache-ssl-tls-mutual-authentication/master/docker-compose.yml)


![](images/iconfinder_note.png)
Il progetto è stato realizzato su macOS Mojave 10.14.4 e testato su Docker Desktop 
CE (versione 2.0.0.3) e Docker (Engine) 18.09.2. Per l'installazione su ambienti
Microsoft Windows consiglio la lettura di [Install Docker Desktop for Windows](https://docs.docker.com/docker-for-windows/install/).

## 1 - Overview
Questo è un progetto [docker](https://www.docker.com/) che parte dall'immagine 
base di [*ubuntu:20.04*](https://hub.docker.com/_/ubuntu), specializzato per 
soddisfare i requisiti minimi per un sistema di mutua autenticazione SSL/TLS.

Il software di base installato è:

* Apache HTTP 2.4 (2.4.41-1ubuntu1)
* PHP 7.4
* PHP 7.4 FPM (FastCGI Process Manager)

_L'installazione di PHP e del modulo per Apache è del tutto opzionale_. I due 
moduli sono stati installati esclusivamente per costruire la pagina di atterraggio
dell'utente dopo la fase di autenticazione. Questa pagina mostra una serie 
d'informazioni base estratte dal certificato digitale utilizzato per 
l'autenticazione.

La mutua autenticazione basata sul protocollo SSL/TLS si riferisce a due parti 
che si autenticano reciprocamente attraverso la verifica del certificato 
digitale fornito in modo che entrambe i partecipanti siano sicuri 
dell'identità altrui.

Da un punto di vista di alto livello, il processo di autenticazione e creazione 
di un canale crittografato utilizzando l'autenticazione reciproca (o mutua 
autenticazione) basata su certificati prevede i passaggi seguenti:

1. Un client richiede l'accesso a una risorsa protetta;
2. Il server presenta il suo certificato al client;
3. Il client verifica il certificato del server;
4. Se ha successo, il client invia il suo certificato al server;
5. Il server verifica le credenziali del cliente;
6. In caso di esito positivo, il server concede l'accesso alla risorsa protetta richiesta dal client.

In Figura 1 è mostrato quello che accade durante il processo di autenticazione 
reciproca (o mutua autenticazione).

![Cosa succede durante il processo di autenticazione reciproca](images/security-sslBMAWithCertificates.gif)

**Figura 1 - Cosa succede durante il processo di autenticazione reciproca (immagine da https://docs.oracle.com)**


## 2 - Struttura del Docker File
Cerchiamo di capire quali sono le sezioni più significative del Dockerfile. 
La prima riga del file (come anticipato in precedenza) fa in modo che il 
container parta dall'immagine docker *ubuntu:20.04*.

```docker
FROM ubuntu:20.04
```

A seguire c'è la sezione delle variabili di ambiente che sono prettamente 
specifiche di Apache HTTP. I valori di queste variabili d'ambiente possono
essere modificate in base alle proprie esigenze.

```docker
# General Apache ENVs
ENV APACHE_RUN_USER www-data
ENV APACHE_RUN_GROUP www-data
ENV APACHE_SERVER_NAME tls-auth.dontesta.it
ENV APACHE_SERVER_ADMIN tls-auth@dontesta.it
ENV APACHE_SSL_CERTS tls-auth.dontesta.it.cer 
ENV APACHE_SSL_PRIVATE tls-auth.dontesta.it.key
ENV APACHE_SSL_PORT 10443
ENV APACHE_LOG_LEVEL info
ENV APACHE_SSL_LOG_LEVEL info
ENV APACHE_SSL_VERIFY_CLIENT optional
ENV APACHE_SSL_SSL_PROXY_ENGINE Off
ENV APACHE_SSL_PROXY_CHECK_PEER_NAME On
ENV APACHE_SERVER_SIGNATURE Off
ENV APACHE_SERVER_TOKENS Prod
ENV APACHE_HTTP_HEADER_X_POWERED_BY "Apache HTTP 2.4 for SSL/TLS Mutual Authentication (Ver. ${VERSION} - Git URL: ${VCS_URL} - Git Ref: ${VCS_REF})"

# For more info See https://httpd.apache.org/docs/2.4/mod/mod_http2.html
ENV APACHE_HTTP_PROTOCOLS http/1.1

# Specifics env Apache for application 
ENV APPLICATION_URL https://${APACHE_SERVER_NAME}:${APACHE_SSL_PORT}
ENV CLIENT_VERIFY_LANDING_PAGE /error.php

# Reverse Proxy Application
ENV APACHE_PROXY_PRESERVE_HOST On
ENV API_BASE_PATH /secure/api
ENV API_BACKEND_BASE_URL http://127.0.0.1:8000${API_BASE_PATH}
```

Il primo gruppo di quattro variabili sono molto esplicative e non necessitano 
approfondimenti. Le variabili a seguire e in particolare 
`APACHE_SSL_CERTS` e `APACHE_SSL_PRIVATE` impostano:

1. il nome del file che contiene il certificato pubblico del server in formato PEM;
2. il nome del file che contiene la chiave privata (in formato PEM) del certificato pubblico.

Il certificato server utilizzato in questo progetto è stato rilasciato da una
_Certification Authority_ privata create ad hoc e ovviamente non riconosciuta. 
Il CN (Common Name) di questo specifico certificato è impostato 
a **tls-auth.dontesta.it** rilasciato dalla _Antonio Musarra's Blog Certification Authority_.

Di default la porta *HTTPS* è impostata a **10443** dalla variabile `APACHE_SSL_PORT`.
La variabile `APPLICATION_URL` definisce il path di redirect qualora si accedesse 
via protocollo HTTP e non HTTPS.

Le variabili `APACHE_LOG_LEVEL`e `APACHE_SSL_LOG_LEVEL`, consentono di modificare
il livello log generale e quello specifico per il modulo SSL. Il valore di default
è impostato a INFO. Per maggiori informazioni potete consultare la documentazione su
[LogLevel Directive](https://httpd.apache.org/docs/2.4/mod/core.html#loglevel).

La variabile `APACHE_SSL_VERIFY_CLIENT` agisce sulla configurazione del processo
di verifica del certificato lato client. Il valore di default è impostato a **optional**.
Rendere opzionale la verifica, consente una gestione più flessibile dell'errore 
in caso che la validazione del certificato client fallisse.

Nel caso in cui il valore della direttiva di Apache **SSLVerifyClient** sia 
**optional** o **optional_no_ca**, se dovesse accadere qualche errore di validazione,
allora sarebbe visualizzata la specifica pagina definita dalla variabile 
`CLIENT_VERIFY_LANDING_PAGE`.

La variabile `APACHE_HTTP_PROTOCOLS` specifica l'elenco di protocolli supportati 
per il server/host virtuale. L'elenco determina i protocolli consentiti che un 
cliente può negoziare per questo server/host.

È necessario impostare i protocolli se si desidera estendere i protocolli 
disponibili per un server/host. Per impostazione predefinita, è consentito solo 
il protocollo **http/1.1** (che include la compatibilità con i client 1.0 e 0.9).

I protocolli validi sono http/1.1 per le connessioni **http** e **https**, 
**h2** sulle connessioni https e **h2c** per le connessioni http.

Per maggiorni informazioni consultare la documentazione 
[Apache Module mod_http2](https://httpd.apache.org/docs/2.4/mod/mod_http2.html).

La sezione a seguire del Dockerfile contiene tutte le direttive necessarie per 
l'installazione del software indicato in precedenza. Dato che la distribuzione scelta 
è [**Ubuntu**](https://www.ubuntu.com/), il comando *apt* è responsabile della gestione 
dei package, quindi dell'installazione.

```docker
# Install services, packages and do cleanup
RUN apt update \
    && apt install -y apache2 \
    && apt install -y php php7.4-fpm \
    && apt install -y curl \
    && apt install -y python3-pip \
    && apt install -y git \
    && rm -rf /var/lib/apt/lists/*
```
 
La sezione a seguire del Dockerfile, anch'essa esplicativa, copia le 
configurazioni di Apache opportunamente modificate al fine di abilitare 
la mutua autenticazione.

```docker
# Copy Apache configuration file
COPY configs/httpd/000-default.conf /etc/apache2/sites-available/
COPY configs/httpd/default-ssl.conf /etc/apache2/sites-available/
COPY configs/httpd/ssl-params.conf /etc/apache2/conf-available/
COPY configs/httpd/dir.conf /etc/apache2/mods-enabled/
COPY configs/httpd/ports.conf /etc/apache2/
```

La sezione a seguire del Dockerfile, copia la coppia di chiavi del server 
e il certificato pubblico della CA.

```docker
# Copy Server (pub and key) tls-auth.dontesta.it
# Copy CA (Certification Authority) Public Key
COPY configs/certs/blog.dontesta.it.ca.cer /etc/ssl/certs/
COPY configs/certs/tls-auth.dontesta.it.cer /etc/ssl/certs/
COPY configs/certs/tls-auth.dontesta.it.key /etc/ssl/private/
``` 

La sezione a seguire del Dockerfile, copia tre script PHP a scopo di test sulla 
*document root* standard di Apache.

```docker
# Copy php samples script and other
COPY configs/www/*.php /var/www/html/
COPY configs/www/assets /var/www/html/assets
COPY configs/www/secure /var/www/html/secure
COPY images/favicon.ico /var/www/html/favicon.ico
```

La sezione a seguire del Dockerfile copia lo script di entrypoint che avvia
server Apache HTTP.

```docker
# Copy scripts and entrypoint
COPY scripts/entrypoint /entrypoint
```

La sezione a seguire del Dockerfile esegue le seguenti principali attività:

1. abilita il modulo SSL
2. abilita il modulo headers
3. abilita il modulo MPM Worker
4. abilita il modulo HTTP2
5. abilita i moduli Proxy, Proxy HTTP e Proxy FCGI (Fast CGI) 
6. abilita il site ssl di default con la configurazione per la mutua autenticazione
7. abilita delle opzioni di configurazione al fine di rafforzare la sicurezza SSL/TLS
8. esegue il re-hash dei certificati. Operazione necessaria affinché Apache sia in grado di leggere i nuovi certificati

```docker
RUN a2enmod ssl \
    && a2enmod headers \
    && a2enmod rewrite \
    && a2dismod mpm_prefork \
    && a2dismod mpm_event \
    && a2enmod mpm_worker \
    && a2enmod proxy_fcgi \
    && a2enmod http2 \
    && a2enmod proxy \
    && a2enmod proxy_http \
    && a2enmod remoteip \
    && a2ensite default-ssl \
    && a2enconf ssl-params \
    && a2enconf php7.4-fpm \
    && c_rehash /etc/ssl/certs/
```

Le due ultime direttive indicate sul Dockerfile, dichiarano la porta HTTPS 
(`APACHE_SSL_PORT`) che deve essere pubblicata e il comando da eseguire per mettere 
in listen (o ascolto) il nuovo servizio Apache HTTP.

## 3 - Organizzazione
In termini di directory e file, il progetto è organizzato così come mostrato a 
seguire. Il cuore di tutto è la directory **configs**.

```
├── Dockerfile
├── configs
│    ├── certs
│    │   ├── blog.dontesta.it.ca.cer
│    │   ├── blog.dontesta.it.ca.key
|    |   ├── tls-auth.dontesta.it.cer
|    |   ├── tls-auth.dontesta.it.key
|    |   ├── tls-auth.dontesta.it.req
|    |   ├── tls-client.dontesta.it.cer
|    |   ├── tls-client.dontesta.it.key
|    |   ├── tls-client.dontesta.it.p12
|    |   ├── tls-client.dontesta.it.req
|    |   ├── mrossi.dontesta.it.cer
|    |   ├── mrossi.dontesta.it.key
|    |   ├── mrossi.dontesta.it.p12
|    |   └── mrossi.dontesta.it.req
│    ├── httpd
│    │   ├── 000-default.conf
│    │   ├── default-ssl.conf
│    │   ├── dir.conf
│    │   ├── ports.conf
│    │   └── ssl-params.conf
│    ├── openssl
│    │   └── openssl.cnf
│    └── wwww
└── scripts
    └── entrypoint
```

La directory *configs* contiene al suo interno altre folder e file, 
in particolare:

1. **certs**
    * contiene il certificato del server (chiave pubblica, chiave privata e CSR);
    * contiene il certificato della CA (chiave pubblica e chiave privata);
    * contiene il certificato client personale per l'autenticazione tramite browser. Sono disponibili: chiave pubblica, chiave privata, CSR, coppia di chiavi in formato PKCS#12;
    * contiene il certificato client da utilizzare per autenticare un'applicazione o dispositivo. Sono disponibili: chiave pubblica, chiave privata, CSR, coppia di chiavi in formato PKCS#12;
2. **openssl**: contiene il file di configurazione per il tool openssl con le impostazioni predefinite;
3. **httpd**: contiene tutte le configurazioni di Apache necessarie per abilitare la mutua autenticazione;
4. **www**: contiene una semplice interfaccia web;
5. **scripts**: contiene l'entrypoint script che avvia il server Apache

## 4 - Quickstart
L'immagine di questo progetto docker è disponibile sul mio account docker hub
[amusarra/apache-ssl-tls-mutual-authentication](https://hub.docker.com/r/amusarra/apache-ssl-tls-mutual-authentication).

A seguire il comando per il **pull** dell'immagine docker ospitata su docker hub. 
Il primo comando esegue il pull dell'ultima versione (tag latest), mentre il 
secondo comando esegue il pull della specifica versione dell'immagine che in 
questo caso è la versione 1.0.0.

```bash
docker pull amusarra/apache-ssl-tls-mutual-authentication
docker pull amusarra/apache-ssl-tls-mutual-authentication:1.0.0
```
Una volta eseguito il pull dell'immagine docker (versione 1.0.0) è possibile 
creare il nuovo container tramite il comando a seguire.

```bash
docker run -i -t -d -p 10443:10443 --name=apache-ssl-tls-mutual-authentication amusarra/apache-ssl-tls-mutual-authentication:1.0.0
```

Utilizzando il comando `docker ps` dovremmo poter vedere in lista il nuovo
container, così come indicato a seguire.

```bash
CONTAINER ID        IMAGE                                  COMMAND                  CREATED             STATUS              PORTS                      NAMES
bb707fb00e89        amusarra/apache-ssl-tls-mutual-authentication:1.0.0   "/usr/sbin/apache2ct…"   6 seconds ago       Up 4 seconds        0.0.0.0:10443->10443/tcp   apache-ssl-tls-mutual-authentication
```

Nel caso in cui vogliate apportare delle modifiche al Dockerfile, dovreste poi 
procedere con la build della nuova immagine e al termine eseguire il run 
dell'immagine ottenuta. A seguire sono indicati i comandi *docker* da eseguire 
dal proprio terminale.

_I comandi docker di build e run devono essere eseguiti dalla root della directory 
di progetto dopo aver fatto il clone di questo repository._

```bash
docker build -t apache-ssl-tls-mutual-authentication .
docker run -i -t -d -p 10443:10443 --name=apache-ssl-tls-mutual-authentication apache-ssl-tls-mutual-authentication:latest
```

A questo punto sul nostro sistema dovremmo avere la nuova immagine con il 
nome **apache-ssl-tls-mutual-authentication** e in esecuzione il nuovo container chiamato
**apache-ssl-tls-mutual-authentication**. Per ragioni di comodità ho chiamato il container
con lo stesso nome dell'immagine, nessuno però vieta di assegnare un nome diverso. 

Utilizzando il comando `docker images` dovremmo poter vedere in lista la nuova
immagine, così come indicato a seguire.

```
REPOSITORY                                      TAG                 IMAGE ID            CREATED             SIZE
apache-ssl-tls-mutual-authentication                           latest              1a145475d1f1        30 minutes ago      242MB
```

Utilizzando il comando `docker ps` dovremmo poter vedere in lista il nuovo
container, così come indicato a seguire.

```
CONTAINER ID        IMAGE                          COMMAND                  CREATED             STATUS              PORTS                      NAMES
65c874216624        apache-ssl-tls-mutual-authentication:latest   "/usr/sbin/apache2ct…"   36 minutes ago      Up 36 minutes       0.0.0.0:10443->10443/tcp   apache-ssl-tls-mutual-authentication
```

Da questo momento è possibile raggiungere il servizio di mutua autenticazione 
SSL/TLS utilizzando il browser. 

Per evitare l'errore `SSL_ERROR_BAD_CERT_DOMAIN` da parte del browser accedendo 
al servizio tramite la URL https://127.0.0.1:10443/, bisogna aggiungere al proprio
file di _hosts_ la riga a seguire.

```
##
# Servizio di mutua autenticazione via Apache HTTPD
##
127.0.0.1       tls-auth.dontesta.it
```

In ambiente di collaudo e/o produzione il nome del servizio o FQDN sarà registrato 
su un DNS.

Lato **server-side** è tutto pronto, manca però una configurazione **client side**,
ovvero, l'installazione del certificato digitale personale sul proprio browser. All'interno
del progetto ho reso disponibili due certificati di esempio che possono essere utilizzati
come certificati client, in particolare:

1. mrossi.dontesta.it.p12
2. tls-client.dontesta.it.p12

Il primo è un certificato digitale _personale_, invece, il secondo certificato digitale può
essere utilizzato per autenticare un'applicazione. Entrambe i certificati (quindi la coppia 
di chiavi) sono contenuti all'interno di un PKCS#12. A seguire sono mostrati i subject dei
rispettivi certificati.

La password di entrambe i PKCS#12 è impostata a: **secret**. Questa è la password
da utilizzare per importare i certificati.

```
Subject: C=IT, L=Rome, ST=Italy, O=Mario Rossi S.r.l, OU=Horse Club, CN=mario.rossi@horseclubsample.com/emailAddress=admin@horseclubsample.com

Subject: C=IT, L=Bronte, ST=Italy, O=Judio Horse Club, OU=IT Services, CN=tls-client.dontesta.it/emailAddress=info@dontesta.it
```

I due certificati sono stati rilasciati dalla CA **Antonio Musarra's Blog Certification Authority**. 
A seguire i dettagli della CA che ha rilasciato tutti i certificati disponibili
all'interno del progetto.

```
Issuer: C=IT, L=Rome, ST=Italy, O=Antonio Musarra's Blog, OU=IT Security Department, CN=Antonio Musarra's Blog Certification Authority/emailAddress=info@dontesta.it
```

Una volta installato il certificato client (file con estensione .p12) sul proprio 
browser (per esempio Firefox), è possibile eseguire il test di mutua autenticazione 
tramite certificato. 

La Figura 2 mostra il certificato digitale installato sul browser Firefox. 
Questo è il certificato da utilizzare per l'autenticazione.

![Certificato Digitale Personale installato sul browser](images/InstalledDigitalPersonalCertified.png)

**Figura 2 - Installazione del certificato digitale personale sul browser Firefox**

Puntando all'indirizzo [https://tls-auth.dontesta.it:10443](https://tls-auth.dontesta.it:10443) dovrebbe accadere quanto segue:

1. Cliccare sul pulsante _Login_ o _LOGIN WITH YOUR DIGITAL CERTIFICATE_
2. Selezionare il certificato digitale installato in precedenza
3. Visualizzazione della pagina di benvenuto.

Le immagini a seguire mostrano il risultato dei tre step precedentemente indicati.

![Welcome Page di Accesso](images/WelcomePageToAccessViaClientCertificate.png)

**Figura 3 - Welcome Page di accesso tramite certificato client**

![Selezione certificato](images/SelectClientCertificate.png)

**Figura 4 - Selezione del certificato client**

![Welcome Page dopo accesso](images/WelcomePageAfterSuccessAuth.png)

**Figura 5 - Welcome Page dopo accesso avvenuto con successo**


Accedendo agli _access log_ di Apache è possibile notare due informazioni utili 
al tracciamento delle operazioni eseguite dall'utente:

* Il protocollo SSL
* Il valore della variabile SSL_CLIENT_S_DN_CN 

```log
172.17.0.1 TLSv1.2 - antonio.musarra@gmail.com [11/Apr/2019:20:17:53 +0000] "GET /secure/ HTTP/1.1" 200 4745 "https://tls-auth.dontesta.it:10443/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:66.0) Gecko/20100101 Firefox/66.0"
```

Il valore di `SSL_CLIENT_S_DN_CN` è inoltre impostato come **SSLUserName**, questo
fa in modo che la variabile `REMOTE_USER` sia impostata con il CN del certificato digitale 
che identifica univocamente l'utente. 

In caso di errore in fase di validazione del certificato client, viene mostrata la pagina
di errore visibile in Figura 6.

![Pagina di errore in caso di certificato non valido](images/CertificateValidationError.png)

**Figura 6 - Pagina di errore in caso di certificato non valido**


## 5 - Build, Run e Push docker image via Makefile
Al fine di semplificare le operazioni di _build_, _run_ e _push_ dell'immagine docker, 

CHANGELOG.md
# Changelog
Tutte le modifiche importanti a questo progetto saranno documentate in questo file.

Il formato è basato su [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
e questo progetto aderisce a [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.8] - 2020-09-17
### Changed
- Updated all SSL/TLS Certificates
- Update Ubuntu image to 20.04

## [1.2.7] - 2020-02-06
### Changed
- Disabled the TLSv1.3 protocol for the issue: TLS 1.3: cannot perform post-handshake authentication.
  You can see this issue at this URL https://bugs.chromium.org/p/chromium/issues/detail?id=911653

## [1.2.6] - 2020-02-06
### Changed
- Updated Ubuntu from 18.04 to 19.10
- Updated Apache HTTP 2.4 to 2.2.41
- Added ServerTokens and ServerSignature Apache HTTP directive
- Added Support for TLS v1.3
- Added Support for Strict Transport Security 

## [1.2.5] - 2020-02-03
### Changed
- Let's Encrypt - Free SSL/TLS Certificates Updated
- Updated - Copyright year

## [1.2.4] - 2019-09-26
### Added
- Added Play-With-Docker button

## [1.2.3] - 2019-09-07
### Added
- Let's Encrypt - Free SSL/TLS Certificates Updated
- Add Apache environment for enable or disable SSL Proxy
- Add Apache environment for enable or disable the remote server certificates
- Add Apache environment for enable or disable proxy preserve host
- Upgrade README

## [1.2.2] - 2019-05-30
### Added
- Add Let's Encrypt - Free SSL/TLS Certificates
- Add the Secure API button
- Upgrade README

## [1.2.1] - 2019-04-25
### Changed
- Upgrade Jinja2 and rope for security reason
- Upgrade README

## [1.2.0] - 2019-04-25
### Added
- Include the [httpbin project](https://github.com/postmanlabs/httpbin.git). Project of [Kenneth Reitz](http://kennethreitz.org/bitcoin)

## [1.1.0] - 2019-04-17
### Added
- Regenerate all certificates with strong SHA2 alg
- Enabled HTTP/2 protocol
- Enabled PHP 7.2 FPM

## [1.0.0] - 2019-04-12
Prima release del progetto. Fare riferimento al README.md per maggiori dettagli
circa il progetto.

configs/www/index.php
<!DOCTYPE html>
<html >
<head>
  <!-- Site made with Mobirise Website Builder v4.8.10, https://mobirise.com -->
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Mobirise v4.8.10, mobirise.com">
  <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
  <link rel="shortcut icon" href="assets/images/iconfinder-21-3319623-1-174x174.png" type="image/x-icon">
  <meta name="description" content="">
  <title>Welcome to Apache SSL/TLS Mutual Authentication Test Environment - Home</title>
  <link rel="stylesheet" href="assets/web/assets/mobirise-icons/mobirise-icons.css">
  <link rel="stylesheet" href="assets/web/assets/mobirise-icons2/mobirise2.css">
  <link rel="stylesheet" href="assets/tether/tether.min.css">
  <link rel="stylesheet" href="assets/bootstrap/css/bootstrap.min.css">
  <link rel="stylesheet" href="assets/bootstrap/css/bootstrap-grid.min.css">
  <link rel="stylesheet" href="assets/bootstrap/css/bootstrap-reboot.min.css">
  <link rel="stylesheet" href="assets/dropdown/css/style.css">
  <link rel="stylesheet" href="assets/socicon/css/styles.css">
  <link rel="stylesheet" href="assets/animatecss/animate.min.css">
  <link rel="stylesheet" href="assets/theme/css/style.css">
  <link rel="stylesheet" href="assets/mobirise/css/mbr-additional.css" type="text/css">
</head>
<body>
  <section class="menu cid-qTkzRZLJNu" once="menu" id="menu1-0">
    <nav class="navbar navbar-expand beta-menu navbar-dropdown align-items-center navbar-toggleable-sm">
        <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
            <div class="hamburger">
                <span></span>
                <span></span>
                <span></span>
                <span></span>
            </div>
        </button>
        <div class="menu-logo">
            <div class="navbar-brand">
                <span class="navbar-logo">
                    <a href="<?= getenv('REQUEST_SCHEME') . '://' . getenv('SERVER_NAME') . ':' . getenv('SERVER_PORT') ?>">
                         <img src="assets/images/iconfinder-21-3319623-1-174x174.png" alt="Mobirise" title="" style="height: 5.6rem;">
                    </a>
                </span>
                <span class="navbar-caption-wrap"><a class="navbar-caption text-white display-4" href="<?= getenv('REQUEST_SCHEME') . '://' . getenv('SERVER_NAME') . ':' . getenv('SERVER_PORT') ?>">Apache SSL/TLS Mutual Authentication</a></span>
            </div>
        </div>
        <div class="collapse navbar-collapse" id="navbarSupportedContent">
            <ul class="navbar-nav nav-dropdown" data-app-modern-menu="true"><li class="nav-item">
                    <a class="nav-link link text-white display-4" href="https://mobirise.com">
                        </a>
                </li>
                <li class="nav-item">
                    <a class="nav-link link text-white display-4" href="https://mobirise.com">
                        </a>
                </li><li class="nav-item"><a class="nav-link link text-white display-4" href="https://www.dontesta.it/">
                        
                        Antonio Musarra's Blog</a></li><li class="nav-item"><a class="nav-link link text-white display-4" href="https://httpd.apache.org/">
                        
                        Apache HTTP Server</a></li></ul>
            <div class="navbar-buttons mbr-section-btn"><a class="btn btn-sm btn-primary display-4" href="<?= getenv('REQUEST_SCHEME') . '://' . getenv('SERVER_NAME') . ':' . getenv('SERVER_PORT') ?>/secure"><span class="mobi-mbri mobi-mbri-login mbr-iconfont mbr-iconfont-btn"></span>
                    
                    Login</a></div>
        </div>
    </nav>
</section>

<section class="engine"><a href="https://mobirise.info/e">make a site for free</a></section><section class="cid-qTkA127IK8 mbr-fullscreen mbr-parallax-background" id="header2-1">

    <div class="mbr-overlay" style="opacity: 0.5; background-color: rgb(118, 118, 118);"></div>

    <div class="container align-center">
        <div class="row justify-content-md-center">
            <div class="mbr-white col-md-10">
                <h1 class="mbr-section-title mbr-bold pb-3 mbr-fonts-style display-1">
                    Apache SSL/TLS Mutual Authentication<br></h1>
                <h3 class="mbr-section-subtitle align-center mbr-light pb-3 mbr-fonts-style display-2">Test Environment</h3>
                <p class="mbr-text pb-3 mbr-fonts-style display-7">
                  The goal of this project is to provide a ready-to-use template that implements a system of <strong>mutual SSL/TLS authentication based on Apache HTTP</strong>.
                  <br/><br/>
                  You can use this environment to perform authentication tests using your digital certificate via the browser or through software that implements a client that uses mutual SSL/TLS authentication.
                </p>
                <div class="mbr-section-btn"><a class="btn btn-md btn-primary display-4" href="https://github.com/amusarra/docker-apache-ssl-tls-mutual-authentication">GO TO THE PROJECT</a>
                    <a class="btn btn-md btn-danger display-4" href="<?= getenv('REQUEST_SCHEME') . '://' . getenv('SERVER_NAME') . ':' . getenv('SERVER_PORT') ?>/secure">LOGIN WITH YOUR DIGITAL CERTIFICATE</a></div>
            </div>
        </div>
    </div>
    <div class="mbr-arrow hidden-sm-down" aria-hidden="true">
        <a href="#next">
            <i class="mbri-down mbr-iconfont"></i>
        </a>
    </div>
</section>

<section class="cid-qTkAaeaxX5" id="footer1-2">

    <div class="container">
        <div class="media-container-row content text-white">
            <div class="col-12 col-md-3">
                <div class="media-wrap">
                    <a href="https://www.dontesta.it">
                        <img src="assets/images/iconfinder-21-3319623-1-174x174.png" alt="Apache SSL/TLS Mutual Authentication" title="Apache SSL/TLS Mutual Authentication">
                    </a>
                </div>
            </div>
            <div class="col-12 col-md-3 mbr-fonts-style display-7">
                <h5 class="pb-3">Antonio Musarra's Blog</h5>
                <p class="mbr-text">
                  <a href="https://www.dontesta.it/2011/05/05/importing-ssl-certificates-on-the-java-keystore-jks/">Importing SSL certificates on the Java Keystore (JKS)</a><br>
                  <a href="https://www.dontesta.it/2017/03/07/richiedere-installare-certificato-ssl-apache-karaf/">Come richiedere e installare un certificato SSL su Apache Karaf</a><br>
                  <a href="https://www.dontesta.it/2014/02/13/https-java-client/">HTTPS Java Client e Certificati SSL</a><br>
                  <a href="https://www.dontesta.it/2017/03/02/come-abilitare-https-apache-karaf-pax-web/">Come abilitare HTTPS su Apache Karaf Pax Web</a><br>
                </p>
            </div>
            <div class="col-12 col-md-3 mbr-fonts-style display-7">
                <h5 class="pb-3">Apache HTTP Server Project</h5>
                <p class="mbr-text">
                  <a href="https://httpd.apache.org/docs/2.4/mod/mod_ssl.html">mod_ssl - Apache HTTP Server Version 2.4</a><br>
                  <a href="https://httpd.apache.org/docs/2.4/ssl/">Apache SSL/TLS Encryption</a><br>
                  <a href="https://httpd.apache.org/docs/2.4/ssl/ssl_intro.html">SSL/TLS Strong Encryption: An Introduction</a></p>
            </div>
            <div class="col-12 col-md-3 mbr-fonts-style display-7">
                <h5 class="pb-3">
                    Links
                </h5>
                <p class="mbr-text">
                  <a href="https://www.dontesta.it">Antonio Musarra's Blog</a>&nbsp;<br>
                  <a href="https://github.com/amusarra/docker-apache-ssl-tls-mutual-authentication">GitHub - Docker Apache SSL/TLS Mutual Authentication</a>&nbsp;<br>
                  <a href="https://hub.docker.com/r/amusarra/apache-ssl-tls-mutual-authentication">Docker Image</a></p>
            </div>
        </div>
        <div class="footer-lower">
            <div class="media-container-row">
                <div class="col-sm-12">
                    <hr>
                </div>
            </div>
            <div class="media-container-row mbr-white">
                <div class="col-sm-6 copyright">
                    <p class="mbr-text mbr-fonts-style display-7"><a href="https://www.dontesta.it">
                        © Copyright 2020 Antonio Musarra's Blog - All Rights Reserved
                    </a></p>
                </div>
                <div class="col-md-6">
                    <div class="social-list align-right">
                        <div class="soc-item">
                            <a href="https://twitter.com/antonio_musarra" target="_blank">
                                <span class="mbr-iconfont mbr-iconfont-social socicon-twitter socicon"></span>
                            </a>
                        </div>
                        <div class="soc-item">
                            <a href="https://www.facebook.com/antoniomusarrablog/" target="_blank">
                                <span class="mbr-iconfont mbr-iconfont-social socicon-facebook socicon"></span>
                            </a>
                        </div>
                        <div class="soc-item">
                            <a href="https://www.youtube.com/channel/UC5D3_EtVPbZYUhrUcEK_THA" target="_blank">
                                <span class="mbr-iconfont mbr-iconfont-social socicon-youtube socicon"></span>
                            </a>
                        </div>
                        <div class="soc-item">
                            <a href="https://www.linkedin.com/in/amusarra/" target="_blank">
                                <span class="mbr-iconfont mbr-iconfont-social socicon-linkedin socicon"></span>
                            </a>
                        </div>
                        <div class="soc-item">
                            <a href="https://github.com/amusarra/" target="_blank">
                                <span class="mbr-iconfont mbr-iconfont-social socicon-github socicon"></span>
                            </a>
                        </div>
                        
                        
                    </div>
                </div>
            </div>
        </div>
    </div>
</section>

  <script src="assets/web/assets/jquery/jquery.min.js"></script>
  <script src="assets/popper/popper.min.js"></script>
  <script src="assets/tether/tether.min.js"></script>
  <script src="assets/bootstrap/js/bootstrap.min.js"></script>
  <script src="assets/dropdown/js/script.min.js"></script>
  <script src="assets/viewportchecker/jquery.viewportchecker.js"></script>
  <script src="assets/parallax/jarallax.min.js"></script>
  <script src="assets/smoothscroll/smooth-scroll.js"></script>
  <script src="assets/touchswipe/jquery.touch-swipe.min.js"></script>
  <script src="assets/theme/js/script.js"></script>
  
  
 <div id="scrollToTop" class="scrollToTop mbr-arrow-up"><a style="text-align: center;"><i></i></a></div>
    <input name="animation" type="hidden">
  </body>
</html>